lib/PuppeteerSharp/Helpers/SendAsyncResponseQueue.cs

```diff
@@ -0,0 +1,105 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using PuppeteerSharp.Messaging;
+
+namespace PuppeteerSharp.Helpers
+{
+    /// <summary>
+    /// Provides an async queue for responses for <see cref="CDPSession.SendAsync"/>, so that responses can be handled
+    /// async without risk callers causing a deadlock.
+    /// </summary>
+    /// <remarks>
+    /// See https://github.com/hardkoded/puppeteer-sharp/issues/1354
+    /// </remarks>
+    internal class SendAsyncResponseQueue : IDisposable
```

Comment:
This class is mostly a refactoring of the initial commit I had with `Task.Run(() => ...)`, but also adds unhandled exception handling, and wrangles any pending tasks if the class that is handling the response async is disposed / closed.

Reply:
Let's call ti `AsyncMessageQueue`.